File: analytics-service/src/main/java/com/emergency/analytics/service/AnalyticsService.java
package com.emergency.analytics.service;

import com.emergency.analytics.dto.*;
import com.emergency.analytics.model.AnalyticsEvent;
import com.emergency.analytics.repository.AnalyticsEventRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;

import java.util.*;
import java.util.stream.Collectors;

@Service
public class AnalyticsService {

    private static final Logger log = LoggerFactory.getLogger(AnalyticsService.class);

    private final AnalyticsEventRepository eventRepository;
    private final WebClient webClient;

    @Value("${services.incident-url}")
    private String incidentServiceUrl;

    @Value("${services.dispatch-url}")
    private String dispatchServiceUrl;

    public AnalyticsService(AnalyticsEventRepository eventRepository, WebClient.Builder webClientBuilder) {
        this.eventRepository = eventRepository;
        this.webClient = webClientBuilder.build();
    }

    /**
     * GET /analytics/response-times
     * Returns average response time statistics from analytics events and live incident data.
     */
    public ResponseTimeAnalytics getResponseTimeAnalytics() {
        // Get from local analytics database
        Double avgResponseTime = eventRepository.findAverageResponseTime();
        List<Object[]> byType = eventRepository.countByIncidentType();

        Map<String, Double> avgByType = new HashMap<>();
        // Simple mock averages per type from the events we have
        byType.forEach(row -> avgByType.put((String) row[0], avgResponseTime != null ? avgResponseTime : 0.0));

        // Also try to get live data from incident service
        long totalIncidents = 0;
        long resolvedIncidents = 0;
        long openIncidents = 0;

        try {
            List<Map<String, Object>> allIncidents = webClient.get()
                    .uri(incidentServiceUrl + "/incidents")
                    .retrieve()
                    .bodyToMono(new ParameterizedTypeReference<List<Map<String, Object>>>() {})
                    .block();

            if (allIncidents != null) {
                totalIncidents = allIncidents.size();
                resolvedIncidents = allIncidents.stream()
                        .filter(i -> "RESOLVED".equals(i.get("status")))
                        .count();
                openIncidents = totalIncidents - resolvedIncidents;

                // Calculate actual response times from dispatched incidents
                for (Map<String, Object> incident : allIncidents) {
                    if (incident.get("timestamp") != null && incident.get("dispatchedAt") != null) {
                        // Record analytics event if not already tracked
                        String incidentType = incident.get("incidentType") != null ?
                                incident.get("incidentType").toString() : "UNKNOWN";
                        avgByType.putIfAbsent(incidentType, avgResponseTime != null ? avgResponseTime : 5.0);
                    }
                }
            }
        } catch (Exception e) {
            log.warn("Could not fetch live data from incident service: {}", e.getMessage());
            totalIncidents = eventRepository.count();
        }

        return ResponseTimeAnalytics.builder()
                .averageResponseTimeMinutes(avgResponseTime != null ? avgResponseTime : 0.0)
                .averageByIncidentType(avgByType)
                .totalIncidents(totalIncidents)
                .resolvedIncidents(resolvedIncidents)
                .openIncidents(openIncidents)
                .build();
    }

    /**
     * GET /analytics/incidents-by-region
     * Returns incident statistics broken down by region and type.
     */
    public IncidentsByRegionAnalytics getIncidentsByRegion() {
        List<Object[]> byRegion = eventRepository.countByRegion();
        List<Object[]> byRegionAndType = eventRepository.countByRegionAndIncidentType();

        Map<String, Long> incidentsByRegion = new LinkedHashMap<>();
        byRegion.forEach(row -> incidentsByRegion.put((String) row[0], (Long) row[1]));

        Map<String, Map<String, Long>> byRegionAndTypeMap = new LinkedHashMap<>();
        byRegionAndType.forEach(row -> {
            String region = (String) row[0];
            String type = (String) row[1];
            Long count = (Long) row[2];
            byRegionAndTypeMap.computeIfAbsent(region, k -> new LinkedHashMap<>()).put(type, count);
        });

        // Also fetch from incident service for live data
        try {
            List<Map<String, Object>> allIncidents = webClient.get()
                    .uri(incidentServiceUrl + "/incidents")
                    .retrieve()
                    .bodyToMono(new ParameterizedTypeReference<List<Map<String, Object>>>() {})
                    .block();

            if (allIncidents != null && !allIncidents.isEmpty()) {
                // Group by a region derived from location (simplified: use locationAddress or generate region from coordinates)
                for (Map<String, Object> incident : allIncidents) {
                    String region = incident.get("locationAddress") != null ?
                            incident.get("locationAddress").toString() : "Unknown Region";
                    // Simplify region to first part
                    if (region.contains(",")) {
                        region = region.substring(0, region.indexOf(",")).trim();
                    }
                    incidentsByRegion.merge(region, 1L, Long::sum);

                    String type = incident.get("incidentType") != null ?
                            incident.get("incidentType").toString() : "UNKNOWN";
                    byRegionAndTypeMap.computeIfAbsent(region, k -> new LinkedHashMap<>())
                            .merge(type, 1L, Long::sum);
                }
            }
        } catch (Exception e) {
            log.warn("Could not fetch live data from incident service: {}", e.getMessage());
        }

        List<IncidentsByRegionAnalytics.RegionStat> topRegions = incidentsByRegion.entrySet().stream()
                .sorted(Map.Entry.<String, Long>comparingByValue().reversed())
                .limit(10)
                .map(entry -> IncidentsByRegionAnalytics.RegionStat.builder()
                        .region(entry.getKey())
                        .count(entry.getValue())
                        .build())
                .collect(Collectors.toList());

        long total = incidentsByRegion.values().stream().mapToLong(Long::longValue).sum();

        return IncidentsByRegionAnalytics.builder()
                .incidentsByRegion(incidentsByRegion)
                .incidentsByRegionAndType(byRegionAndTypeMap)
                .totalIncidents(total)
                .topRegions(topRegions)
                .build();
    }

    /**
     * GET /analytics/resource-utilization
     * Returns resource utilization statistics including vehicle deployment and hospital capacity.
     */
    public ResourceUtilizationAnalytics getResourceUtilization() {
        List<Object[]> byResponder = eventRepository.countByResponderType();
        Map<String, Long> deploymentsByType = new LinkedHashMap<>();
        byResponder.forEach(row -> deploymentsByType.put((String) row[0], (Long) row[1]));

        List<Object[]> byIncidentType = eventRepository.countByIncidentType();
        Map<String, Long> incidentsByType = new LinkedHashMap<>();
        byIncidentType.forEach(row -> incidentsByType.put((String) row[0], (Long) row[1]));

        long totalVehicles = 0;
        long activeVehicles = 0;
        long availableVehicles = 0;

        // Fetch vehicle data from dispatch service
        try {
            List<Map<String, Object>> vehicles = webClient.get()
                    .uri(dispatchServiceUrl + "/vehicles")
                    .retrieve()
                    .bodyToMono(new ParameterizedTypeReference<List<Map<String, Object>>>() {})
                    .block();

            if (vehicles != null) {
                totalVehicles = vehicles.size();
                availableVehicles = vehicles.stream()
                        .filter(v -> "AVAILABLE".equals(v.get("status")))
                        .count();
                activeVehicles = totalVehicles - availableVehicles;
            }
        } catch (Exception e) {
            log.warn("Could not fetch vehicle data from dispatch service: {}", e.getMessage());
        }

        // Fetch station/hospital capacity data from incident service
        Map<String, Object> hospitalCapacity = new LinkedHashMap<>();
        try {
            List<Map<String, Object>> stations = webClient.get()
                    .uri(incidentServiceUrl + "/stations/type/HOSPITAL")
                    .retrieve()
                    .bodyToMono(new ParameterizedTypeReference<List<Map<String, Object>>>() {})
                    .block();

            if (stations != null) {
                int totalCapacity = 0;
                int totalOccupancy = 0;
                for (Map<String, Object> station : stations) {
                    if (station.get("capacity") != null) {
                        totalCapacity += ((Number) station.get("capacity")).intValue();
                    }
                    if (station.get("currentOccupancy") != null) {
                        totalOccupancy += ((Number) station.get("currentOccupancy")).intValue();
                    }
                }
                hospitalCapacity.put("totalBeds", totalCapacity);
                hospitalCapacity.put("occupiedBeds", totalOccupancy);
                hospitalCapacity.put("availableBeds", totalCapacity - totalOccupancy);
                hospitalCapacity.put("utilizationPercentage",
                        totalCapacity > 0 ? (totalOccupancy * 100.0 / totalCapacity) : 0.0);
            }
        } catch (Exception e) {
            log.warn("Could not fetch hospital capacity data: {}", e.getMessage());
        }

        return ResourceUtilizationAnalytics.builder()
                .deploymentsByResponderType(deploymentsByType)
                .incidentsByType(incidentsByType)
                .totalVehicles(totalVehicles)
                .activeVehicles(activeVehicles)
                .availableVehicles(availableVehicles)
                .hospitalCapacity(hospitalCapacity)
                .build();
    }

    /**
     * Records an analytics event for tracking.
     */
    public AnalyticsEvent recordEvent(AnalyticsEvent event) {
        return eventRepository.save(event);
    }
}

File: nodejs-migration/analytics-service/src/services/AnalyticsService.ts
import axios, { AxiosInstance } from 'axios';
import { AppDataSource } from '../database/database';
import { AnalyticsEvent, IncidentType } from '../entities/AnalyticsEvent';
import {
  ResponseTimeAnalytics,
  IncidentsByRegionAnalytics,
  ResourceUtilizationAnalytics,
  AnalyticsEventRequest,
} from '../dto/analytics.dto';

export class AnalyticsService {
  private eventRepository = AppDataSource.getRepository(AnalyticsEvent);
  private httpClient: AxiosInstance;
  private incidentServiceUrl: string;
  private dispatchServiceUrl: string;

  constructor() {
    this.incidentServiceUrl = process.env.INCIDENT_SERVICE_URL || 'http://localhost:3002';
    this.dispatchServiceUrl = process.env.DISPATCH_SERVICE_URL || 'http://localhost:3003';
    this.httpClient = axios.create();
  }

  async getResponseTimeAnalytics(): Promise<ResponseTimeAnalytics> {
    try {
      // Get analytics from local database
      const events = await this.eventRepository.find();
      
      let avgResponseTime = 0;
      if (events.length > 0) {
        const totalTime = events.reduce(
          (sum, event) => sum + (event.responseTimeMinutes || 0),
          0
        );
        avgResponseTime = totalTime / events.length;
      }

      // Group by incident type
      const avgByType: Record<string, number> = {};
      const types = events.reduce((acc, event) => {
        if (event.incidentType) {
          if (!acc[event.incidentType]) {
            acc[event.incidentType] = [];
          }
          acc[event.incidentType].push(event);
        }
        return acc;
      }, {} as Record<string, AnalyticsEvent[]>);

      for (const [type, typeEvents] of Object.entries(types)) {
        const typeAvg = typeEvents.reduce(
          (sum, e) => sum + (e.responseTimeMinutes || 0),
          0
        ) / typeEvents.length;
        avgByType[type] = typeAvg;
      }

      // Try to get live data from incident service
      let totalIncidents = events.length;
      let resolvedIncidents = 0;
      let openIncidents = events.length;

      try {
        const response = await this.httpClient.get(`${this.incidentServiceUrl}/incidents`);
        if (response.data && Array.isArray(response.data)) {
          totalIncidents = response.data.length;
          resolvedIncidents = response.data.filter((i: any) => i.status === 'RESOLVED').length;
          openIncidents = totalIncidents - resolvedIncidents;
        }
      } catch (error) {
        console.warn('Could not fetch live data from incident service');
      }

      return {
        averageResponseTimeMinutes: avgResponseTime,
        averageByIncidentType: avgByType,
        totalIncidents,
        resolvedIncidents,
        openIncidents,
      };
    } catch (error) {
      console.error('Error getting response time analytics:', error);
      throw error;
    }
  }

  async getIncidentsByRegion(): Promise<IncidentsByRegionAnalytics> {
    try {
      const events = await this.eventRepository.find();

      // Group by region
      const byRegion: Record<string, number> = {};
      events.forEach((event) => {
        if (event.region) {
          byRegion[event.region] = (byRegion[event.region] || 0) + 1;
        }
      });

      // Group by region and incident type
      const byRegionAndType = events
        .filter((e) => e.region && e.incidentType)
        .map((e) => ({
          region: e.region!,
          incidentType: e.incidentType!,
          count: 1,
        }))
        .reduce((acc, item) => {
          const existing = acc.find(
            (a) => a.region === item.region && a.incidentType === item.incidentType
          );
          if (existing) {
            existing.count++;
          } else {
            acc.push(item);
          }
          return acc;
        }, [] as Array<{ region: string; incidentType: string; count: number }>);

      return {
        incidentsByRegion: byRegion,
        incidentsByRegionAndType: byRegionAndType,
      };
    } catch (error) {
      console.error('Error getting incidents by region:', error);
      throw error;
    }
  }

  async getResourceUtilization(): Promise<ResourceUtilizationAnalytics> {
    try {
      // Try to get live data from other services
      let resourceDeployment: Array<{ stationType: string; totalUnits: number; availableUnits: number; utilizationPercentage: number }> = [];
      let hospitalBedCapacity: Array<{ hospitalName: string; totalCapacity: number; currentOccupancy: number; occupancyPercentage: number }> = [];
      let avgResponseTime = 0;
      let totalIncidentsServiced = 0;

      try {
        // Get vehicle data
        const vehiclesResponse = await this.httpClient.get(
          `${this.dispatchServiceUrl}/vehicles`
        );
        if (vehiclesResponse.data && Array.isArray(vehiclesResponse.data)) {
          const vehicles = vehiclesResponse.data;
          const byType = vehicles.reduce(
            (acc, v: any) => {
              if (!acc[v.vehicleType]) {
                acc[v.vehicleType] = { total: 0, available: 0 };
              }
              acc[v.vehicleType].total++;
              if (v.status === 'AVAILABLE') {
                acc[v.vehicleType].available++;
              }
              return acc;
            },
            {} as Record<string, { total: number; available: number }>
          );

          resourceDeployment = Object.entries(byType).map(([type, data]) => {
            const typedData = data as { total: number; available: number };
            return {
              stationType: type,
              totalUnits: typedData.total,
              availableUnits: typedData.available,
              utilizationPercentage: ((typedData.total - typedData.available) / typedData.total) * 100,
            };
          });
        }

        // Get station data
        const stationsResponse = await this.httpClient.get(
          `${this.incidentServiceUrl}/stations`
        );
        if (stationsResponse.data && Array.isArray(stationsResponse.data)) {
          const stations = stationsResponse.data;
          hospitalBedCapacity = stations
            .filter((s: any) => s.stationType === 'HOSPITAL')
            .map((s: any) => ({
              hospitalName: s.name,
              totalCapacity: s.capacity || 0,
              currentOccupancy: s.currentOccupancy || 0,
              occupancyPercentage:
                ((s.currentOccupancy || 0) / (s.capacity || 1)) * 100,
            }));
        }
      } catch (error) {
        console.warn('Could not fetch live resource data');
      }

      // Get analytics from local database
      const events = await this.eventRepository.find();
      if (events.length > 0) {
        const totalTime = events.reduce(
          (sum, e) => sum + (e.responseTimeMinutes || 0),
          0
        );
        avgResponseTime = totalTime / events.length;
        totalIncidentsServiced = events.length;
      }

      return {
        resourceDeployment,
        hospitalBedCapacity,
        avgResponseTime,
        totalIncidentsServiced,
      };
    } catch (error) {
      console.error('Error getting resource utilization:', error);
      throw error;
    }
  }

  async recordEvent(request: AnalyticsEventRequest): Promise<AnalyticsEvent> {
    try {
      const event = this.eventRepository.create({
        eventType: request.eventType,
        incidentType: request.incidentType as IncidentType,
        region: request.region,
        responseTimeMinutes: request.responseTimeMinutes,
        incidentId: request.incidentId,
        metadata: request.metadata,
      });

      return await this.eventRepository.save(event);
    } catch (error) {
      console.error('Error recording analytics event:', error);
      throw error;
    }
  }
}
